crates/containerd-shim-wasm/src/sandbox/containerd.rs

```diff
@@ -137,23 +452,148 @@ impl Client {
             log::info!("manifest is not in WASM OCI image format");
             return Ok((vec![], platform));
         };
+
         log::info!("found manifest with WASM OCI image format.");
+        // This label is unique across runtimes and version of the shim running
+        // a precompiled component/module will not work across different runtimes or versions
+        let label = precompile_label(T::info());
+        match image.labels.get(&label) {
+            Some(precompile_digest) if T::can_precompile() => {
+                log::info!("found precompiled module in cache");
+                let precompiled = self.read_content(precompile_digest)?;
+                return Ok((
+                    vec![WasmLayer {
+                        config: image_config_descriptor.clone(),
+                        layer: precompiled,
+                    }],
+                    platform,
+                ));
+            }
+            _ => {}
+        }
 
         let layers = manifest
             .layers()
             .iter()
-            .filter(|x| is_wasm_layer(x.media_type(), supported_layer_types))
-            .map(|config| {
-                self.read_content(config.digest()).map(|module| WasmLayer {
-                    config: config.clone(),
-                    layer: module,
-                })
-            })
+            .filter(|x| is_wasm_layer(x.media_type(), T::supported_layers_types()))
+            .map(|config| self.read_content(config.digest()))
             .collect::<Result<Vec<_>>>()?;
+
+        if T::can_precompile() {
+            log::info!("precompiling module");
+            let precompiled = engine.precompile(layers.as_slice())?;
+            let precompiled_content =
+                self.save_content(precompiled.clone(), image_digest.clone(), T::info())?;
```

Comment:
Would this have visible increase on the latency of running a non-cached image?

Reply:
We don't have a great way to measure this currently.  This will only be called on first run, then it will be cached after this.

Reply:
I used `time` to measure the first run vs the second run and I am seeing the following: 

```
time sudo ctr run --rm --runtime=io.containerd.wasmtime.v1 ghcr.io/contain
erd/runwasi/wasi-demo-oci:latest testwasmoci wasi-demo-oci.wasm echo 'hello'
```

first run: `real    0m1.956s`
second run: `real    0m1.235s`

For an image without OCI layers I see `real   0m2.015s.`  (but it doesn't hit this code path)

These times are really slow generally and not sure if it is my machine/debug mode or what.  But I do see the increase in start up time on the second run.

Reply:
the general slowness was related to DEBUG mode. I turned on Release flag and seeing much better times:

first run: `real    0m0.089s`
second run: `real    0m0.068s`

and for image with out OCI layers: `real    0m1.471s`

Reply:
> I guess this is still pretty slow

I am wondering what benchmark were you comparing to when you said its slow?

Reply:
```
time sudo ctr run --rm docker.io/library/hello-world:latest hello-world
```

First run: 0.345s
Second run: 0.258s

Ah I can see now, it is slow compare to running native image like hello-world in `ctr`, albert the obvious fact that I was running it in a different machine.